# Исследование надежности заемщиков

**Описание проекта**

Заказчик — кредитный отдел банка. Нужно разобраться, влияет ли семейное положение и количество детей клиента на факт погашения кредита в срок. Входные данные от банка — статистика о платёжеспособности клиентов.
Результаты исследования будут учтены при построении модели кредитного скоринга — специальной системы, которая оценивает способность потенциального заёмщика вернуть кредит банку.

**Цель исследования** - проверить гипотезу: 
1. семейное положение и количество детей клиента влияют на факт погашения кредита в срок.

**Ход исследования**

Входные данные от банка - статистика о платежеспособности клиентов в файле `data.csv`. 
Известна структура файла:
* children — количество детей в семье
* days_employed — общий трудовой стаж в днях
* dob_years — возраст клиента в годах
* education — уровень образования клиента
* education_id — идентификатор уровня образования
* family_status — семейное положение
* family_status_id — идентификатор семейного положения
* gender — пол клиента
* income_type — тип занятости
* debt — имел ли задолженность по возврату кредитов
* total_income — ежемесячный доход
* purpose — цель получения кредита

О качестве данных ничего неизвестно. Поэтому перед проверкой гипотезы понадобится обзор данных.
Необходимо проверить данные на ошибки и оценить их влияние на исследование. На этапе предобработки найти возможности исправить самые критичные ошибки данных.

Таким образом, исследование пройдёт в три этапа:
 1. Обзор данных.
 2. Предобработка данных.
 3. Проверка гипотезы.

## Необходимо открыть таблицу и изучить общую информацию о данных

**1. Импортируем библиотеку pandas. Считаем данные из csv-файла в датафрейм и сохраним в переменную `data`**

In [3]:
import pandas as pd #импортируем библиотеку pandas
data = pd.read_csv('data.csv') #считываем данные файла в переменную data

**2. Выведем первые 20 строчек датафрейма `data` на экран.**

In [4]:
data.head(20) #выводим первые 20 записей

children  days_employed  dob_years            education  education_id  \
0          1   -8437.673028         42               высшее             0   
1          1   -4024.803754         36              среднее             1   
2          0   -5623.422610         33              Среднее             1   
3          3   -4124.747207         32              среднее             1   
4          0  340266.072047         53              среднее             1   
5          0    -926.185831         27               высшее             0   
6          0   -2879.202052         43               высшее             0   
7          0    -152.779569         50              СРЕДНЕЕ             1   
8          2   -6929.865299         35               ВЫСШЕЕ             0   
9          0   -2188.756445         41              среднее             1   
10         2   -4171.483647         36               высшее             0   
11         0    -792.701887         40              среднее             1   
12         0            NaN         65              среднее             1   
13         0   -1846.641941         54  неоконченное высшее             2   
14         0   -1844.956182         56               высшее             0   
15         1    -972.364419         26              среднее             1   
16         0   -1719.934226         35              среднее             1   
17         0   -2369.999720         33               высшее             0   
18         0  400281.136913         53              среднее             1   
19         0  -10038.818549         48              СРЕДНЕЕ             1   

       family_status  family_status_id gender income_type  debt  \
0    женат / замужем                 0      F   сотрудник     0   
1    женат / замужем                 0      F   сотрудник     0   
2    женат / замужем                 0      M   сотрудник     0   
3    женат / замужем                 0      M   сотрудник     0   
4   гражданский брак                 1      F   пенсионер     0   
5   гражданский брак                 1      M   компаньон     0   
6    женат / замужем                 0      F   компаньон     0   
7    женат / замужем                 0      M   сотрудник     0   
8   гражданский брак                 1      F   сотрудник     0   
9    женат / замужем                 0      M   сотрудник     0   
10   женат / замужем                 0      M   компаньон     0   
11   женат / замужем                 0      F   сотрудник     0   
12  гражданский брак                 1      M   пенсионер     0   
13   женат / замужем                 0      F   сотрудник     0   
14  гражданский брак                 1      F   компаньон     1   
15   женат / замужем                 0      F   сотрудник     0   
16   женат / замужем                 0      F   сотрудник     0   
17  гражданский брак                 1      M   сотрудник     0   
18    вдовец / вдова                 2      F   пенсионер     0   
19         в разводе                 3      F   сотрудник     0   

     total_income                                 purpose  
0   253875.639453                           покупка жилья  
1   112080.014102                 приобретение автомобиля  
2   145885.952297                           покупка жилья  
3   267628.550329              дополнительное образование  
4   158616.077870                         сыграть свадьбу  
5   255763.565419                           покупка жилья  
6   240525.971920                       операции с жильем  
7   135823.934197                             образование  
8    95856.832424                   на проведение свадьбы  
9   144425.938277                 покупка жилья для семьи  
10  113943.491460                    покупка недвижимости  
11   77069.234271       покупка коммерческой недвижимости  
12            NaN                         сыграть свадьбу  
13  130458.228857                 приобретение автомобиля  
14  165127.911772              покупка жилой недвижимости  
15  116820.904450 

**3. Выведем основную информацию о датафрейме с помощью метода `info()`.**

In [5]:
data.info() #выводим общую информацию

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     19351 non-null  float64
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      19351 non-null  float64
 11  purpose           21525 non-null  object 
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


## Предобработка данных

### Удаление пропусков

**4. Выведем количество пропущенных значений для каждого столбца, используя комбинацию двух методов.**

In [6]:
print(data.isna().sum())

children               0
days_employed       2174
dob_years              0
education              0
education_id           0
family_status          0
family_status_id       0
gender                 0
income_type            0
debt                   0
total_income        2174
purpose                0
dtype: int64


**5. В двух столбцах есть пропущенные значения. Один из них — `days_employed`. Пропуски в этом столбце обработаем на следующем этапе. Другой столбец с пропущенными значениями — `total_income` — хранит данные о доходах. На сумму дохода сильнее всего влияет тип занятости, поэтому заполнить пропуски в этом столбце нужно медианным значением по каждому типу из столбца `income_type`. Например, у человека с типом занятости `сотрудник` пропуск в столбце `total_income` должен быть заполнен медианным доходом среди всех записей с тем же типом.**

In [7]:
data['total_income'] = data.groupby('income_type')['total_income'].transform(lambda g: g.fillna(g.median()))

### Обработка аномальных значений

**6. В данных могут встречаться артефакты (аномалии) — значения, которые не отражают действительность и появились по какой-то ошибке. Таким артефактом является отрицательное количество дней трудового стажа в столбце `days_employed`. Обработаем значения в этом столбце: заменим все отрицательные значения положительными с помощью метода `abs()`.**

In [8]:
data['days_employed'] = data['days_employed'].abs()

**7. Для каждого типа занятости выведем медианное значение трудового стажа `days_employed` в днях.**

In [9]:
print(data.groupby('income_type')['days_employed'].median()) 

income_type
безработный        366413.652744
в декрете            3296.759962
госслужащий          2689.368353
компаньон            1547.382223
пенсионер          365213.306266
предприниматель       520.848083
сотрудник            1574.202821
студент               578.751554
Name: days_employed, dtype: float64


У двух типов (безработные и пенсионеры) получатся аномально большие значения.

**8. Выведем перечень уникальных значений столбца `children`.**

In [10]:
print(data['children'].unique())

[ 1  0  3  2 -1  4 20  5]


**9. В столбце children есть два аномальных значения. Удалим строки, в которых встречаются такие аномальные значения из датафрейма data.**

In [11]:
data = data[(data['children'] >= 0) & (data['children'] < 18)]

**10. Ещё раз выведем перечень уникальных значений столбца `children`, чтобы убедиться, что артефакты удалены.**

In [12]:
print(data['children'].unique())

[1 0 3 2 4 5]


### Удаление пропусков (продолжение)

**11. Заполним пропуски в столбце `days_employed` медианными значениями по каждому типу занятости `income_type`.**

In [13]:
median_days=data.groupby('income_type')['days_employed'].transform('median') 
data['days_employed'] = data['days_employed'].fillna(median_days)

**12. Убедимся, что все пропуски заполнены. Выведем количество пропущенных значений для каждого столбца с помощью двух методов.**

In [14]:
print(data.isna().sum())

children            0
days_employed       0
dob_years           0
education           0
education_id        0
family_status       0
family_status_id    0
gender              0
income_type         0
debt                0
total_income        0
purpose             0
dtype: int64


### Изменение типов данных

**13. Заменим вещественный тип данных в столбце `total_income` на целочисленный с помощью метода `astype()`.**

In [15]:
data['total_income']=data['total_income'].astype('int')

### Обработка дубликатов

**14. Обработаем неявные дубликаты в столбце `education`. В этом столбце есть одни и те же значения, но записанные по-разному: с использованием заглавных и строчных букв. Приведим их к нижнему регистру.**

In [16]:
data['education'] = data['education'].str.lower()

**15. Выведем на экран количество строк-дубликатов в данных. Если такие строки присутствуют, удалим их.**

In [17]:
print(data.duplicated().sum())

71


In [18]:
data = data.drop_duplicates()

### Категоризация данных

**16. На основании диапазонов, указанных ниже, создайте в датафрейме `data` столбец `total_income_category` с категориями:**

- 0–30000 — `'E'`;
- 30001–50000 — `'D'`;
- 50001–200000 — `'C'`;
- 200001–1000000 — `'B'`;
- 1000001 и выше — `'A'`.


**Например, кредитополучателю с доходом 25000 нужно назначить категорию `'E'`, а клиенту, получающему 235000, — `'B'`. Используйте собственную функцию с именем `categorize_income()` и метод `apply()`.**

In [19]:
def categorize_income(total_income):
    if total_income<= 30000:
        return 'E'
    elif total_income<=50000:
        return 'D'
    elif total_income<=200000:
        return 'C'
    elif total_income<=1000000:
        return 'B'
    else:
        return 'A'

In [20]:
data['total_income_category'] = data['total_income'].apply(categorize_income) # применим функцию методом apply()

**17. Выведем на экран перечень уникальных целей взятия кредита из столбца `purpose`.**

In [21]:
print(data['purpose'].unique())

['покупка жилья' 'приобретение автомобиля' 'дополнительное образование'
 'сыграть свадьбу' 'операции с жильем' 'образование'
 'на проведение свадьбы' 'покупка жилья для семьи' 'покупка недвижимости'
 'покупка коммерческой недвижимости' 'покупка жилой недвижимости'
 'строительство собственной недвижимости' 'недвижимость'
 'строительство недвижимости' 'на покупку подержанного автомобиля'
 'на покупку своего автомобиля' 'операции с коммерческой недвижимостью'
 'строительство жилой недвижимости' 'жилье'
 'операции со своей недвижимостью' 'автомобили' 'заняться образованием'
 'сделка с подержанным автомобилем' 'получение образования' 'автомобиль'
 'свадьба' 'получение дополнительного образования' 'покупка своего жилья'
 'операции с недвижимостью' 'получение высшего образования'
 'свой автомобиль' 'сделка с автомобилем' 'профильное образование'
 'высшее образование' 'покупка жилья для сдачи' 'на покупку автомобиля'
 'ремонт жилью' 'заняться высшим образованием']


**Задание 18. Создим функцию, которая на основании данных из столбца `purpose` сформирует новый столбец `purpose_category`, в который войдут следующие категории:**

- `'операции с автомобилем'`,
- `'операции с недвижимостью'`,
- `'проведение свадьбы'`,
- `'получение образования'`.

**Например, если в столбце `purpose` находится подстрока `'на покупку автомобиля'`, то в столбце `purpose_category` должна появиться строка `'операции с автомобилем'`.**

**Используем  функцию с именем `categorize_purpose()` и метод `apply()`. Изучим данные в столбце `purpose` и определим, какие подстроки помогут правильно определить категорию.**

In [22]:
def categorize_purpose(purpose):
    if purpose.find('авто')!=-1:
        return 'операции с автомобилем'
    elif purpose.find('свадьб')!=-1:
        return 'проведение свадьбы'
    elif purpose.find('образов')!=-1:
        return 'получение образования'
    else:
        return 'операции с недвижимостью'

In [23]:
data['purpose_category'] = data['purpose'].apply(categorize_purpose)# применим функцию методом apply()

### Шаг 3. Исследуем данные и ответим на вопросы

#### 3.1 Есть ли зависимость между количеством детей и возвратом кредита в срок?

In [25]:
import numpy as np
data1 = data.pivot_table(index = ['children'], values='debt', aggfunc = [len, sum, np.mean])
data1 = data1.rename(columns={'len': 'Кол-во записей', 'sum': 'Кол-во просрочек', 'mean': 'Доля'})
print(data1)

         Кол-во записей Кол-во просрочек      Доля
                   debt             debt      debt
children                                          
0                 14091             1063  0.075438
1                  4808              444  0.092346
2                  2052              194  0.094542
3                   330               27  0.081818
4                    41                4  0.097561
5                     9                0  0.000000


**Вывод:** Среди заемщиков, у которых есть просрочка, самые "безответственные" (по убыванию) семьи с 4, 2 и 1 ребенком. Более надежными заемщиками являются семьи с 3 детьми - доля просрочек 0.082. 
Cамые надежные заемщики с самой низкой долей просрочек по кредиту 0.075 - это заемщики без детей.
Про заемщиков с 5ю детьми сложно сделать вывод о платежеспособности, т.к. выборка мала и на основании данных по семьям с 4мя детьми можно сделать вывод, что вероятность просрочки будет высокой.

#### 3.2 Есть ли зависимость между семейным положением и возвратом кредита в срок?

In [26]:
data1 = data.pivot_table(index = ['family_status'], values='debt', aggfunc = [len, sum, np.mean])
data1 = data1.rename(columns={'len': 'Кол-во записей', 'sum': 'Кол-во просрочек', 'mean': 'Доля'})
print(data1)

                      Кол-во записей Кол-во просрочек      Доля
                                debt             debt      debt
family_status                                                  
Не женат / не замужем           2796              273  0.097639
в разводе                       1189               84  0.070648
вдовец / вдова                   951               63  0.066246
гражданский брак                4134              385  0.093130
женат / замужем                12261              927  0.075606


**Вывод:** Самый большой процент просрочек по платежам у холостых или находящихся в гражданском (неофициальном) браке заемщиков. У семейных и разведенных заемщиков процент просрочек ниже, но риски высокие. Самыми платежспособными/дисциплинированными заемщиками являются овдовевшие граждане.

#### 3.3 Есть ли зависимость между уровнем дохода и возвратом кредита в срок?

In [27]:
data1 = data.pivot_table(index = ['total_income_category'], values='debt', aggfunc = [len, sum, np.mean])
data1 = data1.rename(columns={'len': 'Кол-во записей', 'sum': 'Кол-во просрочек', 'mean': 'Доля'})
print(data1)

                      Кол-во записей Кол-во просрочек      Доля
                                debt             debt      debt
total_income_category                                          
A                                 25                2  0.080000
B                               5014              354  0.070602
C                              15921             1353  0.084982
D                                349               21  0.060172
E                                 22                2  0.090909


**Вывод:** Самый высокий процент просрочек по платежам у заемщиков с категорией Е (<=30000). Самыми платежспособными/дисциплинированными являеются заемщики с категорией D (с уровнем дохода от 50000 до 200000)

#### 3.4 Как разные цели кредита влияют на его возврат в срок?

In [28]:
data1 = data.pivot_table(index = ['purpose_category'], values='debt', aggfunc = [len, sum, np.mean])
data1 = data1.rename(columns={"len": "Кол-во записей", "sum": "Кол-во просрочек", "mean": "Доля"})
print(data1)

                         Кол-во записей Кол-во просрочек      Доля
                                   debt             debt      debt
purpose_category                                                  
операции с автомобилем             4279              400  0.093480
операции с недвижимостью          10751              780  0.072551
получение образования              3988              369  0.092528
проведение свадьбы                 2313              183  0.079118


**Вывод:** Самый высокий процент просрочек по платежам у заемщиков с целью кредита - операции с автомобилем и чуть ниже для целей получения образования. Самый низкий процент просрочек по кредитам с целью "операции с недвижимостью".

#### 3.5 Возможные причины появления пропусков в исходных данных.

*Ответ:* Основные возможные причины пропусков в исходных данных: проблемы с выгрузкой, непредоставление данных заемщиком. И если обобщить, основными причинами пропусков являются человеческий фактор и техническая ошибка.

#### 3.6 Почему заполнить пропуски медианным значением — лучшее решение для количественных переменных.

*Ответ:* потому что медианное значение - это такое значение, что ровно половина элементов больше него, а другая половина — меньше. Если заполнять средним значением, то данные могут быть искажены (в зависимости от значений). Медиана в таком случае не подвержена влиянию выбросов

### Шаг 4: общий вывод.

#### Цель исследования
Проверить гипотезу: семейное положение и количество детей клиента влияют на факт погашения кредита в срок.

#### Предобработка данных
В рамках проведенного *исследования надежности заемщиков* был проведен анализ полученных данных.
Выявленные проблемы в данных и способы их устранения представлены ниже:

+ **Проблема:** Пропущенные значения количественных переменных; **Решение**: Заполнение медианным значением
+ **Проблема:** Дубли; **Решение**: Приведение к нижнему регистру в строковых переменных, удаления явных дублей
+ **Проблема:** Аномалии данных; **Решение**: Удаление строк с аномальными значениями
+ **Проблема:** Некорректные типы данных; **Решение**: Приведение к нужному типу

Также для анализа потребовалась категоризация данных по следующим критериям: доход, цель кредита.

#### Результаты исследования

Для достижения цели исследования и оценки благонадежности заемщиков мы рассмотрели взаимосвязь наличия просроченных платежей с параметрами: количество детей, семейное положение, цель кредита. 
На основании проведенного исследования можно сделать вывод, что все категории заемщиков имеют зависимость с выплатой платежей по кредиту в срок.

Ниже представлена информация по долям просрочек по категориям.

__Семейное положение__                                                 
* Не женат / не замужем -   **9.7** %
* гражданский брак      -   9.3 %
* женат / замужем       -   7.6 %
* в разводе             -   7.1 %
* вдовец / вдова        -   6.6 %


__Категория дохода__                                          
* 0–30000 — 'E'              -     **9.1** %
* 50001–200000 — 'C'         -     8.5 %
* 1000001 и выше — 'A'       -     8.0 %
* 200001–1000000 — 'B'       -     7.0 %
* 30001–50000 — 'D'          -     6.0 %


__Цель кредита__                                                  
* операции с автомобилем             -  **9.4** %
* получение образования              -  9.3 %
* проведение свадьбы                 -  7.9 %
* операции с недвижимостью           -  7.6 %


Следовательно, самыми надежными заемщиками могут быть:
* 1) без детей
* 2) уровень дохода не более 30000
* 3) с целью кредитования - операции с недвижимостью.

#### Рекомендации

Для полного исследования можно оценить связь с регионом (влияет на уровень дохода и иногда на количество детей), возраст, общее кол-во иждивенцев у заемщика (в данном исследовании рассматривались только дети), а в операциях с недвижимостью - наличие в собственности другой недвижимости.